security/pkg/k8s/chiron/utils.go

```diff
@@ -207,30 +208,35 @@ func submitCSR(wc *WebhookController, csrName string, csrPEM []byte, numRetries
 }
 
 // Read the signed certificate
-func readSignedCertificate(wc *WebhookController, csrName string,
-	readInterval time.Duration, maxNumRead int) ([]byte, []byte, error) {
-	var reqSigned *cert.CertificateSigningRequest
-	for i := 0; i < maxNumRead; i++ {
-		// It takes some time for certificate to be ready, so wait first.
-		time.Sleep(readInterval)
-		r, err := wc.certClient.CertificateSigningRequests().Get(csrName, metav1.GetOptions{})
-		if err != nil {
-			log.Errorf("failed to get the CSR (%v): %v", csrName, err)
-			errCsr := wc.cleanUpCertGen(csrName)
-			if errCsr != nil {
-				log.Errorf("failed to clean up CSR (%v): %v", csrName, err)
+func readSignedCertificate(certClient certclient.CertificateSigningRequestInterface, csrName string,
+	readInterval time.Duration, maxNumRead int, caCertPath string) ([]byte, []byte, error) {
+	// First try to read the signed CSR through a watching mechanism
+	reqSigned := readSignedCsr(certClient, csrName, timeoutForReadingCSR)
+	if reqSigned == nil {
+		// If watching fails, retry reading the signed CSR a few times after waiting.
+		for i := 0; i < maxNumRead; i++ {
+			// It takes some time for certificate to be ready, so wait first.
+			time.Sleep(readInterval)
+			r, err := certClient.Get(csrName, metav1.GetOptions{})
```

Comment:
Just curious: can you explain why Get() is used here and Watch() is used on line #332? In what case, if the CSR response is ready, the Watch() can't obtain it while Get() can obtain it?

Reply:
Get() is used here for simplicity and it takes some time for certificate to be ready, so wait then read. Watch() can also achieve this too, but here Get() is simpler.

Reply:
I see. Then do we really need this? Can we just give a greater value of timeoutForReadingCSR?

Reply:
Giving it a greater value should also work. But here retry mechanism is used just in case Get() failed for transient errors.

Reply:
Not talking about the retries for Get(). I meant the entire block 215 - 235 can be removed if we just give a greater value for timeoutForReadingCSR, right?

Reply:
The block 215-235 is just in case that the watch fails. If you want to remove it, please let me know. I am fine either way.

Reply:
watch() immediately after submitting sometimes return the following error because the CSR is unknown yet. Thus, the code here has the logic "// If watching fails, retry reading the signed CSR a few times after waiting."

err when watching CSR domain-cluster.local-ns-istio-system-secret-dns.istio-galley-service-account: unknown (get certificatesigningrequests.certificates.k8s.io)